test/integration/test_results.py

```diff
@@ -54,6 +54,39 @@ def result_callback(job_id, result):
         self.assertFalse(callback_err)
         self.assertIsNotNone(job._ws_client._server_close_code)
 
+    @run_integration_test
+    def test_result_callback_with_job_result(self, service):
+        """Test result callback along with job result."""
+
+        def result_callback(job_id, result):
+            nonlocal count
+            count = count + 1
+            nonlocal final_it
+            if "iteration" in result:
+                final_it = result["iteration"]
+            nonlocal callback_err
+            if job_id != job.job_id:
+                callback_err.append(f"Unexpected job ID: {job_id}")
+            if "interim_results" in result and result["interim_results"] != int_res:
+                callback_err.append(f"Unexpected interim result: {result}")
```

Comment:
We should consider refactoring this to avoid the need for code duplication here. Similar code is duplicated 4-5 times in this file only.

But I think this is out-of-scope for this PR. Verified that this new test case really fails without your fix - good :).